File: data-structure/stack/stack.js
const stack = {
    _head: null,

    create(data) {
        return {
            next: null,
            data: data
        };
    },

    deleteStack() {
        while (stack._head) {
            const next = stack._head.next;
            delete stack._head;
            stack._head = next;
        }
    },

    pop() {
        let item;

        if ((item = stack._head)) {
            stack._head = item.next;
            item.next = null;
            return item;
        }

        return false;
    },

    push(data) {
        const newItem = stack.create(data);

        if (!stack._head) {
            stack._head = newItem;
        } else {
            newItem.next = stack._head;
            stack._head = newItem;
        }

        return newItem;
    }
};

stack.push(5);
stack.push(7);
stack.push(42);
console.log(stack.pop());
stack.deleteStack();

File: c/data_structures/stack.c
#include <stdio.h>
#include <stdlib.h>

struct node {
    struct node *next;
    void *data;
};

int push(struct node **stack, void *data) {
    struct node *newNode;

    if (!(newNode = malloc(sizeof(struct node)))) {
        fprintf(stderr, "Could not allocate memory for new node!");
        exit(1);
    }

    if (!*stack) {
        newNode->next = NULL;
    } else {
        newNode->next = *stack;
    }

    newNode->data = data;
    *stack = newNode;
}

struct node *pop(struct node **stack) {
    struct node *popped = *stack;

    *stack = popped->next;
    free(popped);

    return popped;
}

int main(void) {
    struct node *stack = NULL;

    push(&stack, "hello, world");
    push(&stack, (int *) 45);
    printf("%d\n", stack->data);

    pop(&stack);
    printf("%s\n", stack->data);

    pop(&stack);
    printf("%p\n", stack);

    return 0;
}

